### PR TITLE
Fix xxe injection vulnerability

### PR DESCRIPTION
**Pull request by haby0**
The Validator class is not used correctly, resulting in an xxe injection vulnerability.

For questions, please contact me.

### CODE REVIEW

**Comment by mariofusco** [comment]
@haby0 can you please clarify what this vulnerability is and how your pull request fixes it?

**Comment by haby0** [comment]
> @haby0 can you please clarify what this vulnerability is and how your pull request fixes it?

When using ```Validator.validate(...)``` to parse xml data, because external entities are not disabled, the external entities are executable. Executing external entities can cause ssrf, dos, and read server files.

You can refer to: https://portswigger.net/web-security/xxe

**Comment by mareknovotny** [comment]
I guess there should be used at least constants defined by jaxp https://docs.oracle.com/javase/tutorial/jaxp/properties/properties.html

- javax.xml.XMLConstants.ACCESS_EXTERNAL_DTD
- javax.xml.XMLConstants.ACCESS_EXTERNAL_SCHEMA
- javax.xml.XMLConstants.ACCESS_EXTERNAL_STYLESHEET

**Comment by tarilabs** [comment]
@mariofusco you may ref https://github.com/kiegroup/drools/pull/2415 for something similar, I did act on the SonarCloud that I did received at the time and applied the suggested corrections (as reported by SonarCloud).
I believe they refer to the same security vulnerability indeed.

**Comment by haby0** [comment]
> I guess there should be used at least constants defined by jaxp https://docs.oracle.com/javase/tutorial/jaxp/properties/properties.html
> 
> * javax.xml.XMLConstants.ACCESS_EXTERNAL_DTD
> * javax.xml.XMLConstants.ACCESS_EXTERNAL_SCHEMA
> * javax.xml.XMLConstants.ACCESS_EXTERNAL_STYLESHEET

You are right.

File: drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java [hunk 1]
```diff
@@ -185,12 +186,18 @@ private static void validate(String kModuleString) {
 
         private static void validate(Source source, Source duplicateSource) {
             try {
-                schema.newValidator().validate(source);
+                Validator validator = schema.newValidator();
+                validator.setProperty("http://javax.xml.XMLConstants/property/accessExternalDTD", "");
```
Comment (mariofusco) [inline drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java:190]: Please replace the hardcoded string "http://javax.xml.XMLConstants/property/accessExternalDTD" with constant `javax.xml.XMLConstants.ACCESS_EXTERNAL_DTD`

File: drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java [hunk 2]
```diff
@@ -185,12 +186,18 @@ private static void validate(String kModuleString) {
...
+                validator.setProperty("http://javax.xml.XMLConstants/property/accessExternalSchema", "");
```
Comment (mariofusco) [inline drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java:191]: Please replace the hardcoded string "http://javax.xml.XMLConstants/property/accessExternalDTD" with constant `javax.xml.XMLConstants.ACCESS_EXTERNAL_SCHEMA`

File: drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java [hunk 3]
```diff
@@ -185,12 +186,18 @@ private static void validate(String kModuleString) {
...
+                validator.validate(source);
             } catch (Exception schemaException) {
                 try {
                     // For backwards compatibility, validate against the old namespace (which has 6.0.0 hardcoded)
                     if (oldSchema != null) {
-                        oldSchema.newValidator().validate( duplicateSource );
+                        Validator oldValidator = oldSchema.newValidator();
+                        oldValidator.setProperty("http://javax.xml.XMLConstants/property/accessExternalDTD", "");
```
Comment (mariofusco) [inline drools-compiler/src/main/java/org/drools/compiler/kproject/models/KieModuleMarshaller.java:198]: Same as above.

**Comment by haby0** [comment]
@mariofusco @tarilabs @mareknovotny Please see the new pr: https://github.com/kiegroup/drools/pull/3808.

**Pull request closed without merging**